### ISSUE TITLE
CocoDetection wrapped with wrap_dataset_for_transforms_v2 returns image_id even if not in target_keys

### ISSUE THREAD

**Issue by rumpg** [pytorch/vision#9162]
### 🐛 Describe the bug

The COCO dataset contains examples without any object. For these samples, the target list returned by CocoDetection is empty. After wrapping the dataset with wrap_dataset_for_transforms_v2, it always returns {'image_id': image_id} for these empty samples, even if "image_id" is not in the target_keys wrapper argument. This can cause bugs downstream, for example when batching, or in custom target transformers relying on the absence of the image_id entry.

See below for a minimal example using a target transformer to encode class ids to multi-label binary vectors:

```
import pathlib

import torch
from torchvision import datasets
from torchvision.transforms import v2

DATA_ROOT = pathlib.Path("data") / "COCO"
ANNOTATIONS_DIR = DATA_ROOT / "annotations"

val_dataset_notransform = datasets.CocoDetection(
    str(DATA_ROOT / "val2017"),
    ANNOTATIONS_DIR / "instances_val2017.json",
)
cat_ids = torch.tensor(sorted(val_dataset_notransform.coco.getCatIds()))
multi_label_binarizer = v2.Lambda(
    lambda y: torch.zeros(
        len(cat_ids), dtype=torch.float
    ).scatter_(
        dim=0, index=torch.bucketize(y, cat_ids), value=1
    )
)

val_dataset = datasets.CocoDetection(
    str(DATA_ROOT / "val2017"),
    ANNOTATIONS_DIR / "instances_val2017.json",
    target_transform=multi_label_binarizer,
)
val_dataset = datasets.wrap_dataset_for_transforms_v2(val_dataset, target_keys=['labels'])

for i in range(len(val_dataset)):
    try:
        val_dataset[i]
    except:
        print(f"Error at {i=}")
        raise
```

### Versions

Collecting environment information...
PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.9 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 18:56:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-122-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.61
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.7.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               60
On-line CPU(s) list:                  0-59
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 6226R CPU @ 2.90GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   1
Core(s) per socket:                   1
Socket(s):                            60
Stepping:                             7
BogoMIPS:                             5786.40
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni md_clear arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.9 MiB (60 instances)
L1i cache:                            1.9 MiB (60 instances)
L2 cache:                             240 MiB (60 instances)
L3 cache:                             960 MiB (60 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-59
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==2.3.0
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.1
[pip3] torchaudio==2.7.1
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] numpy                     2.3.0                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.2                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
[conda] torch                     2.7.1                    pypi_0    pypi
[conda] torchaudio                2.7.1                    pypi_0    pypi
[conda] torchvision               0.22.1                   pypi_0    pypi
[conda] triton                    3.3.1                    pypi_0    pypi

**Comment by AntoineSimoulin** [comment]
Hi @rumpg, Thanks for posting. It seems to me the fact that returns `{'image_id': image_id}` is returned for empty targets is probably due to this [line](https://github.com/pytorch/vision/blob/main/torchvision/tv_tensors/_dataset_wrapper.py#L377). However, in your case, maybe it would be possible to solve your issue by adding some custom logic in your `multi_label_binarizer` transform to control the input type? If not, do you have propositions on how to modify the behavior of the [`coco_dectection_wrapper_factory`](https://github.com/pytorch/vision/blob/main/torchvision/tv_tensors/_dataset_wrapper.py#L340C5-L340C36) to avoid this issue?

**Comment by rumpg** [comment]
Hi @AntoineSimoulin, thanks for your reply.

I could solve my personal issue by subclassing the CocoDetection dataset instead of using transforms.

I think {'image_id': image_id} is the right thing to return when image_id is in target_keys, otherwise it's best to return an empty dictionary instead.
I can contribute a fix in the coming days, when I have the time to setup the development workflow.

**Comment by YeonwooSung** [comment]
I'll work on this: honor `target_keys` for empty CocoDetection samples so `image_id` is only returned when requested.

**Comment by YeonwooSung** [comment]
Opened a fix in https://github.com/pytorch/vision/pull/9650
